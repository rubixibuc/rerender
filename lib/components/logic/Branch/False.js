import React from "react";
import PropTypes from "prop-types";

const render = ({ children }) => <React.Fragment>{children}</React.Fragment>;

render.propTypes = {};

export default render;
