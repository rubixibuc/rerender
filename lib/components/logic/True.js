import React from "react";
import PropTypes from "prop-types";

const render = ({ children }) => <React.Fragment>{children}</React.Fragment>;

render.propTypes = {
  children: PropTypes.node
};

export default render;
