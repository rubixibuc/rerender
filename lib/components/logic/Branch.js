import React from "react";
import PropTypes from "prop-types";

import False from "./False";
import True from "./True";

const render = ({ children, condition }) => {
  const result = typeof condition === "function" ? condition() : condition;

  return (
    <>
      {React.Children.map(
        children,
        child =>
          (result && child.type === True) || (!result && child.type === False)
            ? child
            : null
      )}
    </>
  );
};

const TrueFalsePropType = (
  props,
  propName,
  componentName,
  location,
  propFullName
) =>
  props[propName].type !== True && props[propName].type !== False
    ? new Error(
        "Invalid prop `" +
          propFullName +
          "` supplied to" +
          " `" +
          componentName +
          "`. Validation failed."
      )
    : undefined;

render.propTypes = {
  children: PropTypes.oneOfType([
    TrueFalsePropType,
    PropTypes.arrayOf(TrueFalsePropType)
  ]).isRequired,
  condition: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]).isRequired
};

render.defaultProps = {};

render.displayName = "Rerender.Branch";

export default render;
