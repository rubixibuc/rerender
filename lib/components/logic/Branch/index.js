import React from "react";
import PropTypes from "prop-types";

import False from "./False";
import True from "./True";

const render = ({ children, condition }) =>
  children
    ? React.Children.map(
        children,
        child =>
          (condition() && child.type === True) ||
          (!condition() && child.type === False)
            ? child
            : null
      )
    : null;

const TrueFalsePropType = (props, key) =>
  props[key].type === True || props[key].type === False;

render.propTypes = {
  children: PropTypes.oneOfType([
    TrueFalsePropType,
    PropTypes.arrayOf(TrueFalsePropType)
  ]),
  condition: PropTypes.func.isRequired
};

export default render;
