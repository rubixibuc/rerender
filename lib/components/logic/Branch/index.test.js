import React from "react";
import renderer from "react-test-renderer";

import Branch from "./index";
import False from "./False";
import True from "./True";

test("should render nothing when Branch contains no True or False children and condition is true", () => {
  const component = renderer.create(<Branch condition={() => true} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render nothing when Branch contains no True or False children and condition is false", () => {
  const component = renderer.create(<Branch condition={() => false} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render nothing when Branch contains one True child and condition is false", () => {
  const component = renderer.create(
    <Branch condition={() => false}>
      <True>true</True>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render nothing when Branch contains one False child and condition is true", () => {
  const component = renderer.create(
    <Branch condition={() => true}>
      <False>false</False>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render all false child components when condition is false", () => {
  const component = renderer.create(
    <Branch condition={() => false}>
      <True>
        <div>true</div>
        <div>true</div>
      </True>
      <True>
        <div>true</div>
        <div>true</div>
      </True>
      <False>
        <div>false</div>
        <div>false</div>
      </False>
      <False>
        <div>false</div>
        <div>false</div>
      </False>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("should render all True child components when condition is true", () => {
  const component = renderer.create(
    <Branch condition={() => true}>
      <True>
        <div>true</div>
        <div>true</div>
      </True>
      <True>
        <div>true</div>
        <div>true</div>
      </True>
      <False>
        <div>false</div>
        <div>false</div>
      </False>
      <False>
        <div>false</div>
        <div>false</div>
      </False>
    </Branch>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
