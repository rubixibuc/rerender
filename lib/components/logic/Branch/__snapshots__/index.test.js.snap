// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render all True child components when condition is true 1`] = `
Array [
  <div>
    true
  </div>,
  <div>
    true
  </div>,
  <div>
    true
  </div>,
  <div>
    true
  </div>,
]
`;

exports[`should render all false child components when condition is false 1`] = `
Array [
  <div>
    false
  </div>,
  <div>
    false
  </div>,
  <div>
    false
  </div>,
  <div>
    false
  </div>,
]
`;

exports[`should render nothing when Branch contains no True or False children and condition is false 1`] = `null`;

exports[`should render nothing when Branch contains no True or False children and condition is true 1`] = `null`;

exports[`should render nothing when Branch contains one False child and condition is true 1`] = `null`;

exports[`should render nothing when Branch contains one True child and condition is false 1`] = `null`;
